libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,169 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libsolidity/lsp/ReferenceCollector.h>
+#include <libsolidity/lsp/Utils.h>
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/lsp/LanguageServer.h>
+
+#include <libsolutil/Common.h>
+
+#include <typeinfo>
+
+using namespace solidity::frontend;
+using namespace solidity::langutil;
+using namespace std::string_literals;
+using namespace std;
+
+namespace solidity::lsp
+{
+
+ReferenceCollector::ReferenceCollector(
+	Declaration const& _declaration,
+	string const& _identifierAtCursorLocation
+):
+	m_declaration{_declaration},
+	m_identifierAtCursorLocation{_identifierAtCursorLocation.empty() ? _declaration.name() : _identifierAtCursorLocation}
+{
+}
+
+vector<Reference> ReferenceCollector::collect(
+	Declaration const* _declaration,
+	ASTNode const& _astSearchRoot,
+	string const& _identifierAtCursorLocation
+)
+{
+	if (!_declaration)
+		return {};
+
+	ReferenceCollector collector(*_declaration, _identifierAtCursorLocation);
+	_astSearchRoot.accept(collector);
+	return std::move(collector.m_collectedReferences);
+}
+
+vector<Reference> ReferenceCollector::collect(
+	ASTNode const* _sourceNode,
+	int _sourceOffset,
+	SourceUnit const& _sourceUnit
+)
+{
+	if (!_sourceNode)
+		return {};
+
+	solAssert(_sourceNode->location().containsOffset(_sourceOffset), "");
+
+	auto references = vector<Reference>{};
+
+	if (auto const* identifier = dynamic_cast<Identifier const*>(_sourceNode))
+		references += collect(identifier->annotation().referencedDeclaration, _sourceUnit, identifier->name());
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(_sourceNode))
+	{
+		solAssert(identifierPath->path().size() >= 1, "");
+		for (size_t i = 0; i < identifierPath->pathLocations().size(); ++i)
+		{
+			SourceLocation const location = identifierPath->pathLocations()[i];
+			if (location.containsOffset(_sourceOffset))
+			{
+				Declaration const* declaration = identifierPath->annotation().pathDeclarations.at(i);
+				ASTString const& name = identifierPath->path().at(i);
+				return collect(declaration, _sourceUnit, name);
+			}
+		}
+	}
+	else if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(_sourceNode))
+		references += collect(memberAccess->annotation().referencedDeclaration, _sourceUnit, memberAccess->memberName());
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(_sourceNode))
+		references += collect(declaration, _sourceUnit, declaration->name());
+	else
+		lspRequire(false, ErrorCode::InternalError, "Unhandled AST node "s + typeid(*_sourceNode).name());
+
+	return references;
+}
+
+bool ReferenceCollector::visit(ImportDirective const& _import)
+{
+	for (ImportDirective::SymbolAlias const& symbolAlias: _import.symbolAliases())
+	{
+		if (
+			symbolAlias.alias && *symbolAlias.alias == m_identifierAtCursorLocation &&
```

Comment:
Isn't `symbolAlias.symbol->annotation().referencedDeclaration == &m_declaration` exactly that check?

Reply:
Ah, you're right. I misread it as separate checks for symbol and alias joined with `||`. But it's `&&`. My bad.

But then I see another problem. I think it won't collect `f` in `import {f} from "a.sol";` because it does not have an alias. `f` would normally be visited separately as `IdentifierPath` but in import it won't because this function returns `false` to prevent the visitor from descending deeper.

That's probably also the reason why the module alias isn't collected either as I pointed out in https://github.com/ethereum/solidity/pull/12790/files#r982687552.